core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -800,6 +835,103 @@ private void updateStageExecutions()
+                // already started, nothing to change for subPlan and its descendants
+                return subPlan;
+            }
+
+            if (planFragment.getPartitionCount().isEmpty()) {
+                // no partitioning, then no need to insert extra exchanges to sources
+                return new SubPlan(
+                        planFragment,
+                        subPlan.getChildren().stream().map(child -> overridePartitionCountRecursively(child, oldPartitionCount, newPartitionCount)).collect(toImmutableList()));
+            }
+
+            if (subPlan.getChildren().stream().noneMatch(child -> stageExecutions.containsKey(getStageId(child.getFragment().getId())))) {
+                // none of the source stages have started, then no need to insert extra exchanges to sources
+                return new SubPlan(
+                        planFragment.withPartitionCount(Optional.of(newPartitionCount)),
+                        subPlan.getChildren().stream().map(child -> overridePartitionCountRecursively(child, oldPartitionCount, newPartitionCount)).collect(toImmutableList()));
+            }
+
+            List<RemoteSourceNode> remoteSourceNodes = subPlan.getFragment().getRemoteSourceNodes();
+            ImmutableList.Builder<SubPlan> newSources = ImmutableList.builder();
+            // TODO: the below code inserts extra exchange as the parent of all sources,
+            // can be optimized to those that have started/completed, but directly change the partition count for those who haven't started
+            for (SubPlan source : subPlan.getChildren()) {
+                PlanFragment sourceFragment = source.getFragment();
+                RemoteSourceNode sourceRemoteSourceNode = getOnlyElement(remoteSourceNodes.stream()
+                        .filter(remoteSourceNode -> remoteSourceNode.getSourceFragmentIds().contains(sourceFragment.getId()))
+                        .iterator());
+                requireNonNull(sourceRemoteSourceNode, "sourceRemoteSourceNode is null");
+                RemoteSourceNode runtimeAdaptiveRemoteSourceNode = new RemoteSourceNode(
+                        getRuntimeAdaptivePlanNodeId(sourceRemoteSourceNode.getId()),
+                        sourceFragment.getId(),
+                        sourceFragment.getOutputPartitioningScheme().getOutputLayout(),
+                        sourceRemoteSourceNode.getOrderingScheme(),
+                        sourceRemoteSourceNode.getExchangeType(),
+                        sourceRemoteSourceNode.getRetryPolicy());
+                PlanFragment runtimeAdaptivePlanFragment = new PlanFragment(
+                        getRuntimeAdaptivePlanFragmentId(sourceFragment.getId()),
+                        runtimeAdaptiveRemoteSourceNode,
+                        sourceFragment.getSymbols(),
+                        FIXED_HASH_DISTRIBUTION,
+                        Optional.of(oldPartitionCount),
+                        ImmutableList.of(),
```

Comment:
it is not obvious to me why partitioned sources is always emtpy - can you add explanatory comment in the code?

Reply:
partitioned sources will always be empty --- but remote partitioned sources are not. The input to the `runtimeAdaptiveRemoteSourceNode` will only be from the 'runtimeAdaptiveRemoteSourceNode'